NumPy Operations

In [1]:
import numpy as np
np.set_printoptions(precision=4, suppress=True)
a = np.zeros(24)
b = np.ones((4, 6))
c = np.full((3, 8), 1.0)
a += 1
print("a.dtype =", a.dtype)
print("b.dtype =", b.dtype)
print("c.dtype =", c.dtype)
a = a.reshape((4,6))
c = c.reshape((4,6))
print("a.shape =", a.shape)
print("(a == b).all():", (a == b).all())
print("(b == c).all():", (b == c).all())

a.dtype = float64
b.dtype = float64
c.dtype = float64
a.shape = (4, 6)
(a == b).all(): True
(b == c).all(): True


Ch 1.2

1.2.1 Linear Space

In [2]:
a = np.arange(4)
b = np.ones_like(a)
z = np.zeros((2, 4), np.int64)
m = np.stack([a, b])
m = np.concatenate([m, z]).T
print(m)
np.linalg.eig(m)

[[0 1 0 0]
 [1 1 0 0]
 [2 1 0 0]
 [3 1 0 0]]


(array([ 0.   ,  0.   , -0.618,  1.618]),
 array([[ 0.    ,  0.    , -0.217 , -0.2442],
        [ 0.    ,  0.    ,  0.1341, -0.3952],
        [ 1.    ,  0.    ,  0.4852, -0.5461],
        [ 0.    ,  1.    ,  0.8363, -0.6971]]))

I choose 2 vectors(a, b) randomly here. Eigenvalues of this matrix(m) contains 2 non-zero values (-.618, 1.618) which means those 2 vectors are linear independent. Otherwise(# of non-zero values < 2), they should be considered linear dependent.

1.2.2 Orthogonality

In [3]:
a, b = np.random.rand(4), np.random.rand(4)
print("a =", a, "\nb =", b)
print("inner product of a and b:", np.inner(a, b), sum(a*b))
print("norm of a:", np.linalg.norm(a), np.sqrt(sum(a**2)))

na = a / np.linalg.norm(a)
print("normlized a :", na)
c = (np.dot(b, na))/(np.dot(na, na))*na
print("normlized c :", c / np.linalg.norm(c))
print("c = project b to a:", c)
print("(b-c).c =", np.dot(b-c, c)) # shows that they are orthogonal

a = [0.6291 0.1597 0.1164 0.1887] 
b = [0.4967 0.3333 0.9476 0.1078]
inner product of a and b: 0.49628982385224496 0.49628982385224496
norm of a: 0.6858900499684748 0.6858900499684748
normlized a : [0.9172 0.2328 0.1696 0.2751]
normlized c : [0.9172 0.2328 0.1696 0.2751]
c = project b to a: [0.6637 0.1685 0.1227 0.1991]
(b-c).c = 2.42861286636753e-17


1.2.3 Gram-Schmidt Process

In [4]:
from numpy.core.memmap import dtype
def gram_schmidt(A):
    """Gram-schmidt Orthogonalization"""
    Q=np.zeros_like(A)
    cnt = 0
    for a in A.T:
        u = np.copy(a)
        for i in range(0, cnt):
            u -= np.dot(np.dot(Q[:, i].T, a), Q[:, i]) 
        e = u / np.linalg.norm(u)  # normalization 
        Q[:, cnt] = e
        cnt += 1
    R = np.dot(Q.T, A)
    return (Q, R)

np.set_printoptions(precision=4, suppress=True)
m = np.around(np.random.rand(4, 4)*100)
print("m =", m, "\n")
(Q, R) = gram_schmidt(m)
print("Q =", Q, "\nR =", R)
print("Q.R==m ?", (m-np.dot(Q, R)<0.0001).all())

m = [[48. 64. 96. 12.]
 [90. 86. 81. 25.]
 [66. 88. 68. 61.]
 [65. 23. 43. 34.]] 

Q = [[ 0.3484  0.3654  0.863   0.0165]
 [ 0.6532  0.0025 -0.2511 -0.7144]
 [ 0.479   0.5025 -0.4173  0.5864]
 [ 0.4717 -0.7835  0.1341  0.3815]] 
R = [[137.7861 131.4719 139.2086  65.7686]
 [ -0.      49.8009  35.762    8.4591]
 [  0.       0.      39.9006 -16.8186]
 [  0.       0.       0.      31.0816]]
Q.R==m ? True


1.2.4 Eigenvectors / Eigenvalues

In [5]:
m = np.reshape(np.arange(16)+1, (4, 4))
print("m =", m, "\n")
evalue, evector = np.linalg.eig(m)
print(evalue, "\n", evector, "\n")

for i in range(4):
  print("eigenpair #%d: (a=%.2f, b=%s)\nm*b==a*b ? %s\n" % (i+1, evalue[i], evector[:, i],
                                                        ((np.dot(m, evector[:, i])-evalue[i]*evector[:, i])<0.00001).all()))

m = [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]] 

[36.2094 -2.2094 -0.     -0.    ] 
 [[-0.1512  0.727   0.5037 -0.0646]
 [-0.3492  0.2832 -0.832  -0.3193]
 [-0.5473 -0.1606  0.1528  0.8323]
 [-0.7454 -0.6045  0.1754 -0.4484]] 

eigenpair #1: (a=36.21, b=[-0.1512 -0.3492 -0.5473 -0.7454])
m*b==a*b ? True

eigenpair #2: (a=-2.21, b=[ 0.727   0.2832 -0.1606 -0.6045])
m*b==a*b ? True

eigenpair #3: (a=-0.00, b=[ 0.5037 -0.832   0.1528  0.1754])
m*b==a*b ? True

eigenpair #4: (a=-0.00, b=[-0.0646 -0.3193  0.8323 -0.4484])
m*b==a*b ? True

